File: server/server_main/app/routes/charging.py
from flask import Blueprint, request, jsonify, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from ..extensions import db
from ..models.user import User
from ..models.kiosk import Kiosk
from ..models.charging_session import ChargingSession
from ..services.credit_service import spend_credits
from ..services.command_service import queue_command

charging_bp = Blueprint("charging", __name__)


def _get_current_user() -> User | None:
    user_id = get_jwt_identity()
    return User.query.get(int(user_id))


@charging_bp.post("/start")
@jwt_required()
def start_charging():
    data = request.get_json(silent=True) or {}

    kiosk_id = data.get("kiosk_id")
    port_number = data.get("port_number")
    credits_to_use = int(data.get("credits", 1))

    if not kiosk_id or not port_number:
        return jsonify({"error": "kiosk_id and port_number are required"}), 400

    if port_number not in (1, 2, 3, 4):
        return jsonify({"error": "port_number must be 1-4"}), 400

    kiosk = Kiosk.query.get(kiosk_id)
    if not kiosk:
        return jsonify({"error": "kiosk not found"}), 404

    user = _get_current_user()
    if not user:
        return jsonify({"error": "user not found"}), 404

    minutes_per_credit = current_app.config["MINUTES_PER_CREDIT"]
    duration_seconds = credits_to_use * minutes_per_credit * 60

    # Check port not already active
    active = ChargingSession.query.filter_by(
        kiosk_id=kiosk_id, port_number=port_number, status="active"
    ).first()
    if active:
        return jsonify({"error": f"port {port_number} is already in use"}), 409

    charging = ChargingSession(
        user_id=user.id,
        kiosk_id=kiosk_id,
        port_number=port_number,
        credits_used=credits_to_use,
        duration_seconds=duration_seconds,
    )
    db.session.add(charging)
    db.session.flush()

    spend_credits(user, credits_to_use, "charging_session", charging.id)

    queue_command(
        kiosk_id,
        "activate_port",
        {"port": port_number, "duration_seconds": duration_seconds},
    )

    db.session.commit()

    return jsonify(
        {
            "charging_session": charging.to_dict(),
            "new_balance": user.credit_balance,
        }
    ), 201


@charging_bp.post("/stop/<int:charging_id>")
@jwt_required()
def stop_charging(charging_id):
    user = _get_current_user()
    charging = ChargingSession.query.filter_by(
        id=charging_id, user_id=user.id, status="active"
    ).first()
    if not charging:
        return jsonify({"error": "active charging session not found"}), 404

    charging.interrupt()
    queue_command(charging.kiosk_id, "deactivate_port", {"port": charging.port_number})
    db.session.commit()

    return jsonify({"charging_session": charging.to_dict()})


@charging_bp.get("/active")
@jwt_required()
def active_sessions():
    user = _get_current_user()
    sessions = ChargingSession.query.filter_by(user_id=user.id, status="active").all()
    return jsonify([s.to_dict() for s in sessions])

File: server/server_main/src/routes/charging.ts
import { Router, Response, NextFunction } from 'express'
import { z } from 'zod'
import prisma from '../prisma'
import { requireAuth } from '../middleware/auth'
import { spendCredits } from '../services/creditService'
import { queueCommand } from '../services/commandService'
import { calcDurationSeconds } from '../services/chargingService'
import { AuthRequest } from '../types'

const router = Router()

function sessionToJson(s: {
  id: number
  userId: number
  kioskId: number
  portNumber: number
  creditsUsed: number
  durationSeconds: number
  wattSnapshot: number | null
  startedAt: Date
  endedAt: Date | null
  status: string
}) {
  return {
    id: s.id,
    user_id: s.userId,
    kiosk_id: s.kioskId,
    port_number: s.portNumber,
    credits_used: s.creditsUsed,
    duration_seconds: s.durationSeconds,
    watt_snapshot: s.wattSnapshot,
    started_at: s.startedAt,
    ended_at: s.endedAt,
    status: s.status,
  }
}

const startSchema = z.object({
  kiosk_id: z.number().int().positive(),
  port_number: z.number().int().min(1).max(4),
  credits: z.number().int().positive(),
})

// POST /start
router.post('/start', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const body = startSchema.parse(req.body)
    console.log(`[Charging] START — user #${req.userId} kiosk #${body.kiosk_id} port #${body.port_number} credits=${body.credits}`)

    // Check port not already active
    const activeOnPort = await prisma.chargingSession.findFirst({
      where: { kioskId: body.kiosk_id, portNumber: body.port_number, status: 'active' },
    })
    if (activeOnPort) {
      console.warn(`[Charging] START failed — kiosk #${body.kiosk_id} port #${body.port_number} already in use`)
      res.status(409).json({ error: 'port already in use' }); return
    }

    // Check user has enough credits
    const user = await prisma.user.findUniqueOrThrow({ where: { id: req.userId! } })
    if (user.creditBalance < body.credits) {
      console.warn(`[Charging] START failed — user #${req.userId} insufficient credits (has ${user.creditBalance}, needs ${body.credits})`)
      res.status(400).json({ error: 'insufficient credits' }); return
    }

    // Calculate duration from telemetry
    const { durationSeconds, wattSnapshot } = await calcDurationSeconds(
      body.kiosk_id, body.port_number, body.credits
    )
    console.log(`[Charging] START — duration=${durationSeconds}s watt=${wattSnapshot ?? 0}W`)

    // Create charging session
    const chargingSession = await prisma.chargingSession.create({
      data: {
        userId: req.userId!,
        kioskId: body.kiosk_id,
        portNumber: body.port_number,
        creditsUsed: body.credits,
        durationSeconds,
        wattSnapshot,
        status: 'active',
      },
    })

    // Deduct credits
    const transaction = await spendCredits(req.userId!, body.credits, 'charging_session', chargingSession.id)

    // Queue activate command to ESP32
    await queueCommand(body.kiosk_id, 'activate_port', {
      port: body.port_number,
      duration_seconds: durationSeconds,
    })
    console.log(`[Charging] Session #${chargingSession.id} started — activate_port queued for kiosk #${body.kiosk_id} port #${body.port_number}`)

    res.status(201).json({
      charging_session: sessionToJson(chargingSession),
      new_balance: transaction.balanceAfter,
    })
  } catch (err) {
    next(err)
  }
})

// POST /stop/:id
router.post('/stop/:id', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const id = parseInt(req.params.id)
    console.log(`[Charging] STOP — user #${req.userId} session #${id}`)

    const session = await prisma.chargingSession.findFirst({
      where: { id, userId: req.userId!, status: 'active' },
    })
    if (!session) { console.warn(`[Charging] STOP failed — session #${id} not found or not active for user #${req.userId}`); res.status(404).json({ error: 'active session not found' }); return }

    const updated = await prisma.chargingSession.update({
      where: { id },
      data: { status: 'interrupted', endedAt: new Date() },
    })

    await queueCommand(session.kioskId, 'deactivate_port', { port: session.portNumber })
    console.log(`[Charging] Session #${id} interrupted — deactivate_port queued for kiosk #${session.kioskId} port #${session.portNumber}`)

    res.json(sessionToJson(updated))
  } catch (err) {
    next(err)
  }
})

// GET /active
router.get('/active', requireAuth, async (req: AuthRequest, res: Response, next: NextFunction) => {
  try {
    const sessions = await prisma.chargingSession.findMany({
      where: { userId: req.userId!, status: 'active' },
      orderBy: { startedAt: 'desc' },
    })
    res.json(sessions.map(sessionToJson))
  } catch (err) {
    next(err)
  }
})

export default router
